### PR TITLE
Add token instructions to codecov-action

### LINKED ISSUE #15: Fix codecov token
as can be seen here https://github.com/ros-controls/control_toolbox/pull/188

### PR DESCRIPTION
**Pull request by fmauch**
Since 4v this is required. For this to work, adding a `CODECOV_TOKEN` action secret needs to be added to the GH repository.

This should fix our problems on uploading the code coverage as discussed in https://github.com/ros-controls/control_toolbox/pull/188. Closes #15 

In the target repo a secret has to be added in the project settings:
**
image
**

The token can be extracted on codecov.io.

### CODE REVIEW

**Review by christophfroehlich — COMMENTED** [review]
I already had this, but it didn't work out well.
I tested the `secrets: inherit` option from the calling workflow from control_toolbox repository, but inside my fork I didn't have the token specified obviously. how does this work from our repos as well as from forks?

**Comment by christophfroehlich** [comment]
btw: you can test this with a calling workflow, where you specifiy your repo for the reusable workflow.

**Comment by fmauch** [comment]
I'll play around with that...

File: .github/workflows/reusable-build-coverage.yml [hunk 1]
```diff
@@ -11,6 +11,9 @@ on:
         required: false
         default: 'ubuntu-latest'
         type: string
+    secrets:
+      CODECOV_TOKEN:
+        required: true
```
Comment (christophfroehlich) [inline .github/workflows/reusable-build-coverage.yml:16]: Maybe this is not required if we use `secrets: inherit`?

> Note: If the secrets are inherited by using `secrets: inherit` in the calling workflow, you can reference them even if they are not explicitly defined in the on key.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow

**Comment by christophfroehlich** [comment]
it seems that v4 has lots of issues, and similar reported errors
https://github.com/codecov/codecov-action/issues/1292
should we downgrade it to v4 until it get certainly fixed upstream?

File: .github/workflows/reusable-build-coverage.yml (hunk 1 above)
Comment (fmauch) [reply .github/workflows/reusable-build-coverage.yml:16]: I am not sure whether this is then also unnecessary on the reusable workflow. I'm currently merely testing things, hence the draft state. I'm having the impression that things aren't sorted out with codecov's v4 action, though. With this version I'm having different trouble `Commit creating failed: ["Service not found: none"]` which seems to pop up in https://github.com/codecov/feedback/issues/263. There seems to be a workaround, but I expect that runs us into https://github.com/codecov/codecov-action/issues/1287. There seems to be work around this, e.g. https://github.com/codecov/codecov-action/commit/f62c5eebfe9df5cfe9d56fd86ec925442fe3c214.

We might consider going back to v3 until things are sorted out?

Comment (christophfroehlich) [reply .github/workflows/reusable-build-coverage.yml:16]: I suggested the same in the comment below ;)

**Comment by fmauch** [comment]
Without the custom path I got this working: https://app.codecov.io/gh/fmauch/control_toolbox/tree/test_coverage

File: .github/workflows/reusable-build-coverage.yml [hunk 2]
```diff
@@ -39,12 +42,15 @@ jobs:
...
-      - uses: codecov/codecov-action@v3
+      - uses: codecov/codecov-action@v4
         with:
           fail_ci_if_error: true
           file: ros_ws/lcov/total_coverage.info
           flags: unittests
           name: codecov-umbrella
+          token: ${{ secrets.CODECOV_TOKEN }}
+          verbose: true
+
```
Comment (fmauch) [inline .github/workflows/reusable-build-coverage.yml:53]:
```suggestion
```

**Comment by christophfroehlich** [comment]
Let's check if this works also from forks: https://github.com/fmauch/control_toolbox/pull/1
I configured the other repos with `secrets: inherit`, I'm not sure where the advantages are..

**Comment by fmauch** [comment]
> Let's check if this works also from forks: [fmauch/control_toolbox#1](https://github.com/fmauch/control_toolbox/pull/1) I configured the other repos with `secrets: inherit`, I'm not sure where the advantages are..

https://github.com/fmauch/control_toolbox/actions/runs/8095478742/attempts/1: "Secret CODECOV_TOKEN is required, but not provided while calling."

I can update this PR to not requiring it as an input and see whether inheriting works.

**Comment by fmauch** [comment]
Yep, that seems to be working: https://github.com/fmauch/control_toolbox/actions/runs/8095478742/job/22122490451

**Comment by christophfroehlich** [comment]
Great:

>  info - 2024-02-29 12:01:35,886 -- The PR is happening in a forked repo. Using tokenless upload.

You changed now my PR for ros-controls/control_toolbox. but let's merge this one first. We always can revert it ;)

**Review by christophfroehlich — APPROVED** [review]

**Comment by fmauch** [comment]
> You changed now my PR for ros-controls/control_toolbox. but let's merge this one first. We always can revert it ;)

Ah sorry, I didn't realize you were using the same branch. for creating the PR at my fork.

Regarding the `inherit` vs. explicitly passing in the secret: When passing in the secret explicitly, the workflow could potentially be reused outside of the ros-controls organization. But for the moment I would also be fine with doing it that way.

**Pull request merged**